src/evaluators/ForStatement.js

```diff
@@ -268,13 +280,151 @@ function ForBodyEvaluation(
   }
 }
 
+let BailOutWrapperClosureRefVisitor = {
+  ReferencedIdentifier(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    if (path.node.name === "arguments") {
+      state.usesArguments = true;
+    }
+  },
+  ThisExpression(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    state.usesThis = true;
+  },
+  "BreakStatement|ContinueStatement"(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    if (path.node.label !== undefined) {
+      state.usesGotoToLabel = true;
+    }
+  },
+  ReturnStatement(path: BabelTraversePath, state: BailOutWrapperInfo) {
+    state.usesReturn = true;
+  },
+};
+
+function generateRuntimeForStatement(
+  ast: BabelNodeForStatement,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm,
+  labelSet: ?Array<string>
+): AbstractValue {
+  let wrapperFunction = new ECMAScriptSourceFunctionValue(realm);
+  let body = t.blockStatement([ast]);
+  ((body: any): FunctionBodyAstNode).uniqueOrderedTag = realm.functionBodyUniqueTagSeed++;
+  wrapperFunction.$ECMAScriptCode = body;
+  wrapperFunction.$FormalParameters = [];
+  wrapperFunction.$Environment = env;
+  // We need to scan to AST looking for "this", "return", labels and "arguments"
+  let functionInfo = {
+    usesArguments: false,
+    usesThis: false,
+    usesReturn: false,
+    usesGotoToLabel: false,
+  };
+
+  traverse(
+    t.file(t.program([t.expressionStatement(t.functionExpression(null, [], body))])),
+    BailOutWrapperClosureRefVisitor,
+    null,
+    functionInfo
+  );
+  traverse.clearCache();
+
+  if (functionInfo.usesReturn || functionInfo.usesArguments || functionInfo.usesGotoToLabel) {
+    // We do not have support for these yet
+    // TODO: issue number upon creation of issue
+    throw new FatalError("TODO: #? handle more loop bail out cases");
+  }
+  let args = [wrapperFunction];
+
+  if (functionInfo.usesThis) {
+    let thisRef = env.evaluate(t.thisExpression(), strictCode);
+    let thisVal = Environment.GetValue(realm, thisRef);
+    Havoc.value(realm, thisVal);
+    args.push(thisVal);
+  }
+
+  Havoc.value(realm, wrapperFunction);
```

Comment:
Can you add a comment to explain why the wrapper needs to get havoced? Preferably, there should be a test that fails if it is not havoced.

Reply:
The wrapper references all relevant bindings, which in turn get havoced.

Reply:
Which answers my previous question. It would be good to point this out in the actual code. It is really not that obvious.